lightningd/htlc_set.c

```diff
@@ -0,0 +1,193 @@
+#include <common/memleak.h>
+#include <common/timeout.h>
+#include <lightningd/htlc_end.h>
+#include <lightningd/htlc_set.h>
+#include <lightningd/lightningd.h>
+#include <lightningd/peer_htlcs.h>
+
+/* Pre-declaration as fail_htlc_set refers to it and vice versa */
+static void htlc_set_hin_destroyed(struct htlc_in *hin,
+				   struct htlc_set *set);
+
+void fail_htlc_set(struct htlc_set *set, enum onion_type failcode)
+{
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		/* Don't remove from set */
+		tal_del_destructor2(set->htlcs[i], htlc_set_hin_destroyed, set);
+		fail_htlc(set->htlcs[i], failcode);
+	}
+	tal_free(set);
+}
+
+/* If an HTLC times out, we need to free entire set, since we could be processing
+ * it in invoice.c right now. */
+static void htlc_set_hin_destroyed(struct htlc_in *hin,
+				   struct htlc_set *set)
+{
+	/* For non-EXPERIMENTAL_FEATURES, this never actually called. */
+#if EXPERIMENTAL_FEATURES
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		if (set->htlcs[i] == hin) {
+			/* Don't try to re-fail this HTLC! */
+			tal_arr_remove(&set->htlcs, i);
+			/* Kind of the correct failure code. */
+			fail_htlc_set(set, WIRE_MPP_TIMEOUT);
+			return;
+		}
+	}
+	abort();
+#endif
+}
+
+void fulfill_htlc_set(struct htlc_set *set, const struct preimage *preimage)
+{
+	for (size_t i = 0; i < tal_count(set->htlcs); i++) {
+		/* Don't remove from set */
+		tal_del_destructor2(set->htlcs[i], htlc_set_hin_destroyed, set);
```

Comment:
This is done independently of whether it was set as a destructor, hence the function sig above being outside of the compile guards, correct? Just checked, `tal_del_destructor2` just returns `false` if the function was never registered. Why not make this call and the function definition above experimental only?

Reply:
Yes, that's probably even clearer.